sqlglot/dialects/fabric.py

```diff
@@ -65,6 +65,19 @@ class Generator(TSQL.Generator):
         }
 
         def datatype_sql(self, expression: exp.DataType) -> str:
+            # DATETIMEOFFSET is only supported when used with AT TIME ZONE:
+            # https://learn.microsoft.com/en-us/sql/t-sql/data-types/datetimeoffset-transact-sql#microsoft-fabric-support
+            if expression.is_type(exp.DataType.Type.TIMESTAMPTZ):
+                parent = getattr(expression, "parent", None)
+                grandparent = getattr(parent, "parent", None) if parent else None
+
+                if not (grandparent and isinstance(grandparent, exp.AtTimeZone)):
+                    # Not used with AT TIME ZONE: treat as TIMESTAMP (DATETIME2)
+                    expression = exp.DataType(
+                        this=exp.DataType.Type.TIMESTAMP,
+                        expressions=expression.expressions,
+                    )
```

Comment:
Not sure if this approach is very robust. It assumes that `AtTimeZone` will only appear as a grandparent, which is true for `CAST` expressions, but I'm not sure if it can be generalized to other expressions where a `DataType` may appear.

I'd suggest generalizing it as follows:

```python
attimezone = expression.find_ancestor(exp.AtTimeZone, exp.Select)
if not attimezone:
    expression = exp.DataType(this=exp.DataType.Type.TIMESTAMP, expressions=expression.expressions)
```

Including `Select` in the above argument list is done to prune the walk path at the `SELECT` boundary.